### ISSUE TITLE
Use parent context in slots?

### ISSUE THREAD

**Issue by JuroOravec** [EmilStenstrom/django-components#350]
## Context

In frontend frameworks like Vue, it's normal to expect that when you define slots, e.g.:

```vue
<!-- Parent.vue -->
<MyComponent>
  Hello {{ firstName }}!
</MyComponent>
```

that the variables used INSIDE the slot belong to the parent context. That is, the `firstName` is taken from the file/component, where this markup was written (`Parent.vue`).

Now, if we want django-components to properly separate context of components, we need to use the `only` keyword.

However, as I found today, the problem is that using the `only` keyword also cuts off slot's access to the parent context.

So in the example above, if it was written in Django, the `firstName` would be undefined / empty.

This seem like a design shortcoming - I can either separate contexts, or cannot use parent's context in slots. But not both, like in Vue.

## Workaround

To allow us to use slots with the parent's context in our project, I subclassed django-component's `Component` to pre-render the slot fills (available via `Component.fill_content`), so that, at the point of render, the slot content is no longer dependent on the parent context.:

```py
from django.template.base import Node, NodeList, TextNode
import django_components.component

def prerender_component_slots(comp: django_components.component.Component):
    def prerender_node(node: Node):
        return TextNode(node.render(comp.outer_context))

    # Case: Single default slot
    if isinstance(comp.fill_content, NodeList):
        rendered = prerender_node(comp.fill_content)
        comp.fill_content = NodeList([rendered])

    # Case: Multiple named slots
    # Where each is NamedFillContent - tuple of (slot_name, Node, alias)
    else:
        comp.fill_content = [
            (
                slot_fill[0],
                NodeList([
                    prerender_node(slot_fill[1])
                ]),
                slot_fill[2],
            )
            for slot_fill in comp.fill_content
        ]
```

## Closing thoughts

Can we get something similar into django-components? Has it been considered before? (didn't find relevant issues)

**Comment by EmilStenstrom** [comment]
Hi! Thanks for a well worded issue that both goes into the use-case and also shows some example code!

As an author of a components library there are many decisions like this, when one framework does it one way, and another does it another way. I have no intention of making django-components "like Vue", but if there's a good case for Vue to do it that way, I'm definitely open to doing the same.

So, let's start there. Why is to useful to have access to the component's parents context, but not the full context, when rendering a context?

My thinking around the only keyword is to enforce strict separation, so that global variables get much harder to abuse. It's still rather easy to include them though, you just send them in when calling the component:

```django
{% component_block "card" firstName=firstName only %}
   ...
{% endcomponent_block %}
```

Could you give some more background and use-cases?

**Comment by JuroOravec** [comment]
Hey @EmilStenstrom, thanks for the swift reply!

To understand why I think using the parent's slot is preferred in this case, you have to consider where and how the slots are being used.

Tl;DR - It's about 3 things:
- Separation of child and parent logic
- Minimizing unexpected behavior
- Minimizing repetition

To explain, let's reuse the `firstName` example:

```django
<!-- parent.html -->
{% component_block "child" only %}
  Hello {{ first_name }}!
{% endcomponent_block %}
```

And the child component could look like this:

```django
<!-- child.html -->
<div class="{{ color }}">
  {% slot "content" default %} {% endslot %}
<div>
```

```py
# child.py
@component.register("child")
class Child(Component):
    template_name = "child.html"
    def get_context_data(self, bg_color: str):
        return {
            "color": prepare_color(bg_color),
        }
```

## Separation of child and parent logic

Consider that:
1. Without the workaround, `first_name` would not render.
2. We can render `first_name` without the workaround by passing it as a prop to `Child` as you've shown. However, for this, we have to explicitly define `first_name` in `get_context_data`:
	```django
	<!-- parent.html -->
	{% component_block "child" first_name=first_name only %}
	  Hello {{ first_name }}!
	{% endcomponent_block %}
	```

	```py
	# child.py
	@component.register("child")
	class Child(Component):
	    template_name = "child.html"
	    def get_context_data(self, bg_color: str, first_name: str):
	        return {
	            "color": prepare_color(bg_color),
				"first_name": first_name,
	        }
	```

3. But what happens if we use the component in another place, with a different content? We have to update the child component again...

	```django
	<!-- parent2.html -->
	{% component_block "child" last_name=last_name request=request only %}
	  Hello {{ last_name }}! Your request ID is {{ request.id }}.
	{% endcomponent_block %}
	```

	```py
	# child.py
	@component.register("child")
	class Child(Component):
	    template_name = "child.html"
	    def get_context_data(self, bg_color: str, first_name: str, last_name: str, request):
	        return {
	            "color": prepare_color(bg_color),
				"first_name": first_name,
				"last_name": last_name,
				"request", request,
	        }
	```

	This becomes cumbersome - every time we want an extra variable inside the slot, we have to define it as an input and pass it to the context.

4. Instead of manually defining the `first_name`, `last_name`, `request`, we can alleviate things by allowing to pass any kwargs to the child:

	```py
	# child.py
	@component.register("child")
	class Child(Component):
	    template_name = "child.html"
	    def get_context_data(self, bg_color: str, **kwargs):
	        return {
	            "color": prepare_color(bg_color),
				**kwargs,
	        }
	```

	Oops, but we actually have to be careful to put kwargs *before* any of our props, to avoid user accidentally overriding them:
	```py
    return {
	    **kwargs,
        "color": prepare_color(bg_color),
    }
	```

5. But what if in some case the parent component also uses the `color` variable and there's still conflict?

	```django
	<!-- parent3.html -->
	{% component_block "child" bg_color="red" color="blue" only %}
	  My favorite color is {{ color }}!
	{% endcomponent_block %}
	```

	In this case, we would print `My favorite color is "red"!` instead of `My favorite color is "blue"!`.

	From just looking at the usage, there's no clear indication that `bg_color` is internally converted to `color`.

	So the parent STILL needs to know about the implementation of the child component.

## Minimizing unexpected behavior and repetition

Why is the lack of proper separation an issue?

- It's hard to debug
- Get's repetitive fast.

If I'm trying to debug why I'm getting `red` instead of `blue`, where would I look? In this case, there's only 1 place to look, so it's ok.

But what if we had 3 and more nested components?

Imagine this below is a markup for a page that has a button to delete some "changes". On clicking a button, a confirmation dialog opens. So the button is in a "activator" slot, meaning that when the "activator" is clicked, the dialog is opened. This is not far from how we use django-components in our project:

```django
{% component_block "layout" ... only %}
	{% component_block "dialog" ... only %}
	  {% fill "activator" %}
	    {% component_block "button" ... only %}
	      {% component "icon" icon="trash" ... only %}
	      Delete {{ changes_count }} changes
	    {% endcomponent_block %}
	  {% endfill %}
	{% endcomponent_block %}
{% endcomponent_block %}
```

Notice that we always use `only` as a good coding practice. But thus, as expected, the `changes_count` doesn't render! 

What can I do?

1. Remove the `only` keywords
	- This is the simplest change, not far from ideal. We lose the predictability. It's like defining global variables. Now I cannot add another variable to the context without fearing that I might break something in one of the child components:

	```django
	{% component_block "layout" ... %}
		{% component_block "dialog" ... %}
		  {% fill "activator" %}
		    {% component_block "button" ... %}
		      {% component "icon" icon="trash" ... %}
		      Delete {{ changes_count }} changes
		    {% endcomponent_block %}
		  {% endfill %}
		{% endcomponent_block %}
	{% endcomponent_block %}
	```

2. Pass the props as kwargs
	- Just in this small example, I have to define `changes_count` 6x to pass it through. What if the slot needed 5 variables? I'd be already using half of all the text to just pass the props.

	```django
	{% component_block "layout" changes_count=changes_count ... only %}
		{% component_block "dialog" changes_count=changes_count ... only %}
		  {% fill "activator" %}
		    {% component_block "button" changes_count=changes_count ... only %}
		      {% component "icon" icon="trash" ... only %}
		      Delete {{ changes_count }} changes
		    {% endcomponent_block %}
		  {% endfill %}
		{% endcomponent_block %}
	{% endcomponent_block %}
	```

3. Render slots in parent's context
	- Using parent's context when rendering slots is the ideal solution here - I can still use `only` to have the assurance that I'm not messing up with the child components, yet I don't have to needlessly drill the prop

	```django
	{% component_block "layout" ... only %}
		{% component_block "dialog" ... only %}
		  {% fill "activator" %}
		    {% component_block "button" ... only %}
		      {% component "icon" icon="trash" ... only %}
		      Delete {{ changes_count }} changes
		    {% endcomponent_block %}
		  {% endfill %}
		{% endcomponent_block %}
	{% endcomponent_block %}
	```

## Ok, but what if I really want side effects?

I hope that by now you see the appeal of rendering slots using the parent's context.

Since the opposite is the true right now, I can imagine there can be others could have already built components that intentionally make use of the fact that slots are rendered in the child's context. Maybe something like this:

```django
{% component_block "truncated_card" name=name address=address only %}
  Hello {{ name }}!

  My address is {{ address }}
{% endcomponent_block %}
```

Where the component would intentionally modify the kwargs:

```py
# truncated_card.py
@component.register("truncated_card")
class TruncatedList(Component):
    template_name = "truncated_card.html"
    def get_context_data(self, **kwargs):
        return {
			key: truncate(str(val), 20) for key, val in kwargs
        }
```

And there's nothing wrong with that. Since that's how things have worked so far, that's how it should remain too.

So I think that as a first step, the "using parent's context for slots" feature should be opt-in. For example another keyword on the child component similarly to `only`. E.g. `context`.

So tagging the component with `context` would hint to use parent's context for rendering the slot fills:

- So using `only` AND `context` would render `first_name` fine, because it'd be taken from parent's context:

	```django
	{% component_block "child" only context %}
	  Hello {{ first_name }}!
	{% endcomponent_block %}
	```

- Whereas using `only` WITHOUT `context` would make the slot to be rendered with child's context, where `first_name` is undefined. So this would NOT render `first_name`:

	```django
	{% component_block "child" only %}
	  Hello {{ first_name }}!
	{% endcomponent_block %}
	```

- And as recap, without either `only` OR `context` the `first_name` would rendered using the child's context, BUT `first_name` would be available to ALL components, so this WOULD render `first_name`:

	```django
	{% component_block "child" %}
	  Hello {{ first_name }}!
	{% endcomponent_block %}
	```


---

Let me know what you think, and happy holidays!

**Comment by EmilStenstrom** [comment]
Thanks again for a well-argued case. What I'm still missing is an explaination of what this new "context" mode means. 

If we look at the good real-life example you posted:

```django
{% component_block "layout" ... only %}
  {% component_block "dialog" ... only %}
    {% fill "activator" %}
      {% component_block "button" ... only %}
        {% component "icon" icon="trash" ... only %}
        Delete {{ changes_count }} changes
      {% endcomponent_block %}
    {% endfill %}
  {% endcomponent_block %}
{% endcomponent_block %}
```
... with the assumption that `changes_count` is a variable from the django view context (right?).

I see that only wouldn't work here (and indeed that this is a common use case).

How is this new "context" mode different from not using only at all?

**Comment by JuroOravec** [comment]
In short, when you don't use `only` at all, the components share state. It's as if you had a global state. And it's possible that someone else will edit one of the components and unintentionally overwrite the variable. (see the example below).

With the "context mode", the slot fill can access the parent's template context, but not contexts of other components. So we don't have the issue with the shared state. (slot fill's "parent" is the component for whom we've defined the HTML file).

---

Example of why NOT using `only` is a bad idea:

Imagine we're working in a team.

Let's say you wrote something like this:

```django
{# my_page.html #}
{% component_block "button" %}
  Some data: {{ data }}
{% endcomponent_block %}
```

I, as your teammate, didn't work with your file, so I'm not aware that you used the `data` variable inside the `button` component's slot.

Imagine I want to add a feature to the `button` component to automatically track user activity on button click. I will add a template context variable that holds the telemetry data. And I will name it `data`, because it seems fine in the context of the `button.py` file.

```py
# button.py
@component.register("button")
class Button(Component):
    template_name = "button.html"
    def get_context_data(self, **kwargs):
        data = prepare_tracking_data()
        return {
            "data": data,
			**kwargs,
        }
```

At this point, I accidentally broke your code in `my_page.html`. I would need to know about ALL variables inside ALL slots in ALL `button` instances to make sure that my addition would NOT break something. That's simply too brittle.

That's why NOT using `only` is a bad idea if we're working on more complex projects.

And why I don't think that simply omitting `only` is a sufficient solution here.

**Comment by EmilStenstrom** [comment]
I understand the case for not omitting `only`.

> Parent of the slot fill is the component for whom we've defined the HTML file).

In the last example I've posted. What's the scoping rules like for that component in "context mode"?

**Comment by JuroOravec** [comment]
To explain the rules, I think the best is to return to my implementation. In our project, I subclassed `Component` class, and run the following function at the beginning of `get_context_data`:

<details>
<summary>
prerender_component_slots
</summary>

```py
from django.template.base import Node, NodeList, TextNode
import django_components.component

def prerender_component_slots(comp: django_components.component.Component):
    def prerender_node(node: Node):
        return TextNode(node.render(comp.outer_context))

    # Case: Single default slot
    if isinstance(comp.fill_content, NodeList):
        rendered = prerender_node(comp.fill_content)
        comp.fill_content = NodeList([rendered])

    # Case: Multiple named slots
    # Where each is NamedFillContent - tuple of (slot_name, Node, alias)
    else:
        comp.fill_content = [
            (
                slot_fill[0],
                NodeList([
                    prerender_node(slot_fill[1])
                ]),
                slot_fill[2],
            )
            for slot_fill in comp.fill_content
        ]
```

</details>

Since the function runs inside a method of `Component` class, it can pre-render Component's slots accessible at `self.fill_content` and uses the `self.outer_context` as the context.

Hence, with this implementation, ALL slots defined in the template are pre-rendered with parent's context. So I could do something like this:

```django
{{ my_var }}
{% component_block "layout" ... only %}
  {{ my_var }}
    {% component_block "dialog" ... only %}
    {% fill "activator" %}
      {{ my_var }}
      {% component_block "button" ... only %}
        {{ my_var }}
        {% component "icon" icon="trash" ... only %}
        {{ my_var }}
      {% endcomponent_block %}
    {{ my_var }}
    {% endfill %}
  {% endcomponent_block %}
  {{ my_var }}
{% endcomponent_block %}
{{ my_var }}
```

And if `my_var` was `"hello"`, then it would render:


```django
hello
{% component_block "layout" ... only %}
  hello
    {% component_block "dialog" ... only %}
    {% fill "activator" %}
      hello
      {% component_block "button" ... only %}
        hello
        {% component "icon" icon="trash" ... only %}
        hello
      {% endcomponent_block %}
    hello
    {% endfill %}
  {% endcomponent_block %}
  hello
{% endcomponent_block %}
hello
```

Now... as I'm describing it, I wonder. Since it's either that all slots use the parent's context, or none, then it would make more sense for the flag to be defined on the `Component` subclass instead of the in the template.

So instead of flagging the slot with `context` inside the django template as I suggested earlier:

```django
{% component_block "layout" ... only context %}
    {{ my_var }}
{% endcomponent_block %}
```

Instead, it would be defined in Python on the Component class:

```py
@component.register("button")
class Button(Component):
    template_name = "components/button/button.html"

    class Media:
        css = "components/button/button.css"
        js = "components/button/button.js"

    render_slots_in_parent_context = True

    def get_context_data(self, *args, **kwargs):
        ...
```

(We'd want a snappier name than `render_slots_in_parent_context`, that's just for exaggaration to make it clear how it should work.)

**Comment by EmilStenstrom** [comment]
Thanks again for explaining. I'm warming up to the idea.

In your example were my_var is "hello": Since all components are rendered with in the parent context, this means that all the top level variables are propagated to the innermost level, right? So in that case, are we not back to the same behaviour as with global variables? So if some developer would set my_var to something else, it could break your components? Or is the advantage that a component's own variables takes precedence over global ones?

**Comment by JuroOravec** [comment]
Let's imagine that the example with `my_var="hello"` was a Django template for the `button` component shown in my previous message.

1. If the `button` component itself was rendered WITHOUT the `only` flag, then yes, it would be identical to global variables (same as omitting `only`), so `render_slots_in_parent_context` would have no effect.

    ```django
    {% component "button" %}
    ```

2. But if `button` component was rendered WITH the `only` flag, then:

	```django
	{% component "button" only %}
	```

    - if `render_slots_in_parent_context = False` or not set, then we would have same outcome as is currently.
    - if `render_slots_in_parent_context = True`, then the slot would be rendered only with the context defined in the `button` component

**Comment by EmilStenstrom** [comment]
I think I'm missing something that you see clearly. I think I need an example were I can see how this is beneficial.

In the `{% component_block "layout" ... only %}` case, the new isolation level is the same as skipping "only".
In the `{% component "button" %}` example, it's again the same as skipping "only".

If the intention is to nest components that are all accepting their parent context, I don't see the difference with just skipping "only". So I'm guessing there's some other case I don't see?

(BTW, we have a global setting now, that controls the default isolation level, so you don't have to put only on each component if you want every component to behave a specific way)

**Comment by EmilStenstrom** [comment]
@JuroOravec Did you give up on me? :)

**Comment by JuroOravec** [comment]
Hey @EmilStenstrom, not yet 😄 (Jk), thanks for the bump, I've missed the previous message.

Maybe a fresh take on this - Let's say that if we set `render_slots_in_parent_context = True` on a `Component A`, then ALL the double curly brackets (`{{ }}`) in the template get rendered with the context of `Component A`, irrespective of how deep they are nested.

As for what you wrote in the message from last month, no, those examples are not exactly the same as skipping "only". If I skip "only", then any component between the "root" component and the double curlies can override the value without me knowing. But if I use `render_slots_in_parent_context = True`, then I have the guarantee that the value will be 100% the value I defined  in the "root".

Without `render_slots_in_parent_context = True`:
```django
{{ my_var }}                             <----- value of `my_var` is 1
{% component_block "comp_a" %}           <----- component "comp_a" sets value of `my_var` to 2
  {{ my_var }}
    {% component_block "comp_b" %}       <----- component "comp_a" sets value of `my_var` to 3
      {{ my_var }} 
    {% endfill %}
  {% endcomponent_block %}
{% endcomponent_block %}
```

With `render_slots_in_parent_context = True`:
```django
{{ my_var }}                             <----- value of `my_var` is 1
{% component_block "comp_a" %}           <----- value of `my_var` is 1
  {{ my_var }}
    {% component_block "comp_b" %}       <----- value of `my_var` is 1
      {{ my_var }} 
    {% endfill %}
  {% endcomponent_block %}
{% endcomponent_block %}
```

Also I'm happy to chat/call on this, it's definitely easier to show, then to try to describe.

**Comment by EmilStenstrom** [comment]
@JuroOravec Thanks for giving a complete example, I understand the difference finally :D

I'm not sure I agree that this is the best way to resolve variables. If I understand it correctly all "global variables" (variables set in the root) can't be overridden in the components? I would expect variables to behave like python classes and subclasses, that any variable set in the root CAN be overridden by a nested component, but that the override only applies inside that component. This is the default currently.

Going back to your use-case: Do you currently have an issue with components setting/defining variables that shadow the global ones? I can see how a component authors could accidentally use a variable name that a parent component already uses, and then break the parent. But in your case, now the child component breaks instead, because it can't override the variable it expects to.

I wonder if the correct way to do this is to raise an exception when a component reuses a variable that's defined in the root?

Then again, what can you as a developer if this happens? In the worst case, both the components you want to use are built by a third party, so can't be changed. Then you're just stuck.

Another way to go about this would be to isolate each component's context automatically, renaming all variables in both the context dict and in the template to avoid errors. That would avoid this issue altogether, and allow different components to use the same variable names if they want.

Thoughts?

**Comment by JuroOravec** [comment]
Hey @EmilStenstrom thanks for the reply.

I think that it boils down to the question of what kind of philosophy should be applied in this project.

I also see your point better now that you've mentioned that you see the nested components like subclasses. From that perspective I'd agree that it'd make sense to expect that nested components would have access to outer/parent variable by default.

The perspective where I'm coming from is from Vue and React, where things are different conceptually. In those frameworks, components are not inherited, but composed. So components have strict separation, and (usually) the behaviour of a component could be changed only via inputs.*

> NOTE: Lines are bit more blurry with [React's contexts](https://react.dev/reference/react/hooks#context-hooks) and [Vue's inject/provide](https://vuejs.org/guide/components/provide-inject), which can effectively change the behaviour of a component without changing the component's input.

Focusing now only on the case of filling slots, in both Vue and React, the interaction with variables of a nested components are OPT-IN.

In Vue, you have to use scoped slots to explicitly access the variables that the child component exposed to the slot:

```vue
<div>
  <MyComponent1>
	<!-- Taken from "root" component -->
    {{ myVar }}
  </MyComponent1>
  <MyComponent2 #default="slotProps">
	<!-- Taken from "MyComponent2" -->
    {{ slotProps.myVar }}
  </MyComponent1>
</div>
```

In React, slot fills are passed in as component inputs. Since these slot fills are defined BEFORE they are passed to the nested component, and in their own function scope, the slot fills don't use the variables from the nested component to render the content.

In the example below, if we rendered `ParentComponent`, the content would say "Shh, don't tell!"

```tsx
function ParentComponent(props) {
  const secret = "Shh, don't tell!"
  return (
    <div>
      <MyReactComponent2>
		{secret}
      </MyReactComponent2>
    </div>
  );
}

function MyReactComponent2(props) {
  const secret = "Inner secret"
  return (
     <span>
		{props.children}
     </span>
  );
}
```

If we wanted to really pass variables to nested components in React, we'd have to make the slot fills into functions, so they can be evaluated with the nested component's variables. This is practically the same as [Vue's scoped slots](https://vuejs.org/guide/components/slots.html#scoped-slots). And again it's the parent component that decides whether and how to use the variables from the nested component:

```tsx
function ParentComponent(props) {
  const secret = "Shh, don't tell!"
  const nestedSlotFill = (nestedSecret) => {
    return (
        <span>
          {nestedSecret}
        </span>
    );
  };

  return (
    <div>
      <MyReactComponent2 mySlot={nestedSlotFill} />
    </div>
  );
}

function MyReactComponent2(props) {
  const secret = "Inner secret"
  const renderedFill = props.mySlot(secret);
  return (
     <span>
		{renderedFill}
     </span>
  );
}
```

---

As for my use case, frankly, the chance of shadowing/conflicting variables is low, because we're a small team.

My issue is that it feels like it would be more difficult to write bigger projects with django-components, because there isn't that inherent trust that what you see is what you get.

**Comment by JuroOravec** [comment]
Now that I wrote the above, it got me thinking. Since django-components declares and fills slots similar to Vue, what do you think resolving this by going in the direction of Vue. That is, that:
1. The child component would define which variables it exposes to the slot
2. Parent component opts in to use the child component's variables

So slot definition in the child component could look like so:

```django
<div>
  {% slot "myslot" var_one="hello" var_two=some_var %}
</div>
```

And then the parent component would use the variables exposed by the child like so:

```django
<h1>
  {% component_block "child" %}
    {% fill "myslot" var_one %}
      {{ my_var }}
      {{ var_one }}
      {{ var_two }}
    {% endfill %}
  {% endcomponent_block %}
</h1>
```

So in the example above:
- `my_var` is taken from the parent component
- `var_one` is taken from the child component, because it was exposed to the slot fill.
- `var_two` was not declared in slot fill, so it would be taken from the parent component

**Comment by EmilStenstrom** [comment]
I haven't followed the React and Vue worlds for a long white, thanks for explaining how they work relative to this. I'm now convinced that they are grown from being user-friendly to being friendly-to-very-large-projects instead :D. There's sooo much complexity in them!

About this issue: Reading your last example code I've **finally** understood what has been in the title of this issue all along 🤦 (reread the thread again), that we are talking about component context and slots context **within the same component**, not about nested components per se. I have no idea how I've missed that for so long.

I totally understand the need to pass variables from the parent context to the slots when in "only mode".

I hadn't realized that any component that uses component context in slots stops working when used in "only mode". **The intent of the only tag was to avoid leaking page context into components**, NOT shielding slots from its own component context. Changing this to allow slots to use the parent's context makes more sense to me now. That change should be backwards compatible too, since we're moving to more permissive.

The suggested syntax you write would then not be needed if I understand you correctly? And we would also avoid having components not work in only mode because their variables would suddenly not be available.

**Comment by JuroOravec** [comment]
Haha, yeah, but it's always like that with complexity, isn't it? Like for example the web server template I use (NodeJS), it has 4 layers to it (browser client, server interface (e.g. GraphQL), server operation that handles the request, and database layer through ORM). And at the end of the day, it's just fancy way of building SQL queries and returning the data 😄 But that sort of complexity is good when you want to have control over every stage of the flow.

---

Anyway, as for this issue, seems like we're on the same page now! Thanks for your patience, I'm sure I could've described it better, but I was also just refining the idea as I went along. 🙏 

So yes, to reiterate using your words, this is about "the `only` tag not shielding slots from its own component context".

And yes, so in this case we could do away with any extra syntax.

So in the example below, when we render the `Parent` component, we should see `456`, because the `Parent` component used `only` on `Child` component.

```py
@component.register("child")
class Child(component.Component):
    def get_context_data(self, date):
        return {
            "variable": 123,
        }
    
    template: t.django_html = """
        <div>
          {% slot "content" default %}{% endslot %}
        </div>
    """
```

```py
@component.register("parent")
class Parent(component.Component):
    def get_context_data(self, date):
        return {
            "variable": 456,
        }
    
    template: t.django_html = """
        {% component "child" only %}
          {{ variable }}
        {% endcomponent %}
    """
```

One question remains though. What should be rendered in the example above if we didn't use the `only` tag? Should it be 123, or 456? Based on the Vue and React examples, my take would be that it should still render 456.

**Comment by JuroOravec** [comment]
Addressed in #437
